pkg/nodeagent/controller/healthcheck/kubelet.go

```diff
@@ -0,0 +1,271 @@
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/utils/clock"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/gardener/gardener/pkg/nodeagent/dbus"
+)
+
+const (
+	// kubeletServiceName is the systemd service name of the kubelet
+	kubeletServiceName = "kubelet.service"
+	// defaultKubeletHealthEndpoint is the health endpoint of the kubelet
+	defaultKubeletHealthEndpoint = "http://127.0.0.1:10248/healthz"
+	// maxToggles defines how often the kubelet can chang the readiness during toggleTimeSpan until the node will be hard rebooted
+	maxToggles = 5
+	// toggleTimeSpan is a floating time window where the kubelet readiness toggles are considered harmful
+	toggleTimeSpan = 10 * time.Minute
+)
+
+// KubeletHealthChecker configures the kubelet healthcheck
+type KubeletHealthChecker struct {
+	// Clock exported for testing
+	Clock clock.Clock
+	// KubeletReadynessToggles contains a entry for every toggle between kubelet Ready->NotReady or NotReady->Ready state
+	KubeletReadynessToggles []time.Time
+
+	client   client.Client
+	nodeName string
+	// firstFailure stores the time of the first failed kubelet health check
+	firstFailure *time.Time
+	dbus         dbus.DBus
+	recorder     record.EventRecorder
+	// lastInternalIP stores the node internalIP
+	lastInternalIP netip.Addr
+	// getAddresses is a func which returns a slice of ip addresses on this node
+	getAddresses          func() ([]net.Addr, error)
+	nodeReady             bool
+	kubeletHealthEndpoint string
+}
+
+// NewKubeletHealthChecker create a instance of a kubelet healthcheck
+func NewKubeletHealthChecker(client client.Client, nodeName string, clock clock.Clock, dbus dbus.DBus, recorder record.EventRecorder, getAddresses func() ([]net.Addr, error)) *KubeletHealthChecker {
+	return &KubeletHealthChecker{
+		client:                  client,
+		nodeName:                nodeName,
+		dbus:                    dbus,
+		Clock:                   clock,
+		recorder:                recorder,
+		getAddresses:            getAddresses,
+		KubeletReadynessToggles: []time.Time{},
+		kubeletHealthEndpoint:   defaultKubeletHealthEndpoint,
+	}
+}
+
+// Name returns the name of this healthcheck.
+func (*KubeletHealthChecker) Name() string {
+	return "kubelet"
+}
+
+// HasLastInternalIP returns true if the node.InternalIP was stored
+// exported for testing
+func (k *KubeletHealthChecker) HasLastInternalIP() bool {
+	return k.lastInternalIP.IsValid()
+}
+
+// SetKubeletHealthEndpoint set the kubeletHealthEndpoint
+// exported for testing
+func (k *KubeletHealthChecker) SetKubeletHealthEndpoint(kubeletHealthEndpoint string) {
+	k.kubeletHealthEndpoint = kubeletHealthEndpoint
+}
+
+// Check performs the actual health check for the kubelet.
+func (k *KubeletHealthChecker) Check(ctx context.Context) error {
+	log := logf.FromContext(ctx).WithName("kubelet")
+	node := &corev1.Node{}
```

Comment:
Similar to `containerd` health check this does not do a metadata get. Is that intentional?

Reply:
I moved getting the node from `Check` to `Reconciler` function.
The reason for this is that the initial assumption `nodeName == hostName` is wrong, so we don't know the node name when the controllers are initialized. 
There is already a node predicate we can use, so it is easier to get the node from `Reconcile` method.

Kubelet health check reads the status of the node, so we need to get the entire node object here.